Functional test environment

brybacki opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Add a set of scripts to create test infrastructure. Such scripts are probably easier to manage than current programmatic way (which needs  some additional depencies - like cdi/kubevirt client and so on), because they use standard yaml files and kubectl apply. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This is still a work in progress. A few more steps are needed to finish it:
- finish the script (set hotplug feature flag)
- updated documentation (mentioning https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-kubevirt/main/kubevirtci),
- remove code that creates test cluster 
- create and always run a simple smoke test - requires new test lane
- `bindata` - probably needs to be updated 

After that is done we can finally focus on making sure all functional tests are running on a test lane. 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

phoracek commented on .gitignore:
```diff
@@ -18,6 +18,9 @@ _out/*
 
 # kubevirtCI
 _ci-configs/
+/cluster-up/
```
It seems that you've committed the cluster-up directory though

brybacki replied:
thanks for noticing that

brybacki replied:
No, I have verified it, there is no cluster-up dir. 
What I have commited is a change that removes the cluter-up dir and all files there.

davidvossel commented on kubevirtci:
```diff
@@ -0,0 +1,415 @@
+
+function kubevirtci::build() {
+	export REGISTRY="127.0.0.1:$(cluster-up/cluster-up/cli.sh ports registry)"
+	make docker-build
+	make docker-push
```
you'll want to have `make` targets that build the csi container image and can push it to the kubevirtci repo

brybacki replied:
ok, I need to use these two targets: https://github.com/kubevirt/csi-driver/pull/40

davidvossel reviewed (commented):
good start. I think this logic could be broken up into the following make targets

- `make cluster-up`: brings up a kubevirtci cluster with a capk guest cluster completely installed on it
- `make image-build`:  builds kubevirt csi container image locally
- `make image-push`: pushes the kubevirt csi container image to a repo (defaults to the kubevirtci local repo)
- `make cluser-sync`: installs kubevirt csi yaml into the kubevirtci cluster. The yaml should reference the container image in the local kubevirtci repo (from the `image push` step)

Then eventually

- `make e2e-test`: Executes the e2e tests on the kubevirt ci cluster.

brybacki commented:
Improved this a little bit, and after running and trying to create volumes, it looks like it works correctly. 

So now I need to add make targets, and update the csi-yamls or create a new set of templates (to use images build with make docker-build). And of course I can make the bash code nicer, but first I want to see it working.

davidvossel commented on hack/cluster-sync.sh:
```diff
@@ -0,0 +1,56 @@
+
+./kubevirtci kubectl-tenant create -f ./deploy/000-csi-driver.yaml
+./kubevirtci kubectl-tenant create -f ./deploy/020-autorization.yaml
+./kubevirtci kubectl-tenant create -f ./deploy/030-node.yaml
+./kubevirtci kubectl-tenant create -f ./deploy/040-controller.yaml
```
once the container image is synced into the kubevirtci cluster, these two files will have to be altered. At that point I think we're kind of entering into `kustomize` territory with how these yamls likely need to be handled. This might also help with other yaml that need to be mutated (like perhaps storage class)

It might be worth restructuring the existing yaml into a kustomize directory tree similar to how the capk repo works.
[1] is the kustomize base dir
[2] is the kustomize overlay `kubevirtci` patch set
[3][4] shows how the local image is injected into kustomize's yaml for the kubevirt ci overlay when the images are built
[5] shows how the entire thing is deployed in the kubevirtci script

So for the image part, the workflow could be something like...

`make image-build` which should both build and update the kustomize's kubevirtci overlay to point to the image that the deploy needs to use
`make image-push` which then pushes the image to the kubevirtci repo

That way when you later do a `${_kubectl} kustomize config/kubevirtci | ${_kubectl} apply -f -` the local image is used correctly.


1. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/main/config
2. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/main/config/kubevirtci
3. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/48146e1c41d730c13e159b11d147c9f0f42361e6/Makefile#L233
4. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/main/config/kubevirtci/manager_image_patch.yaml
5. https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/48146e1c41d730c13e159b11d147c9f0f42361e6/kubevirtci#L147

davidvossel commented on hack/cluster-sync.sh:
```diff
@@ -0,0 +1,56 @@
+
+export INFRA_KUBECONFIG_STORAGECLASS_FILE=_ci-configs/storageclass.yaml
+cp ./deploy/example/storageclass.yaml $INFRA_KUBECONFIG_STORAGECLASS_FILE
+sed -i -r 's/standard/rook-ceph-block/g' $INFRA_KUBECONFIG_STORAGECLASS_FILE
```
i can't remember the details here. Does this need to match the storageclass used on the infra cluster? If so, i dont' believe rook-ceph is deployed by kubevirtci anymore. We transitioned to using local storage instead due to issues with resource consumption with rook-ceph.

so maybe use `local` here?

awels replied:
You can enable using rook-ceph (if you want) by calling `export KUBEVIRT_STORAGE="rook-ceph-default"` before calling make cluster-up. This should deploy ceph (its part of the default kubevirtci image)

davidvossel replied:
yep, we're aware. The issue is resource consumption. We're kind of on the edge (memory wise) due to needing to run two giant VMs nested in the environment to host the guest cluster. Unless we need rook-ceph, local storage would be helpful. We actually removed usage of rook-ceph in the capk repo because of memory issues.

awels replied:
I guess local can work, if you want something more dynamic you can try hpp, it doesn't use much memory. But local storage is likely even less memory

davidvossel replied:
fyi, this still isn't resolved, which might be causing issues for you

brybacki replied:
did not see issues, I did tests on both ceph and local, both work, it is now updated, will push soon

brybacki replied:
done - updated to local

davidvossel commented on hack/cluster-sync.sh:
```diff
@@ -0,0 +1,56 @@
+# ******************************************************
+# Create service account and kubeconfig to access tenant cluster
+# ******************************************************
+./kubevirtci kubectl -n kvcluster create -f ./deploy/infra-cluster-service-account.yaml
```
this is kind of a broad suggestion for all this automation.

maybe use `apply` instead of `create`. That way if the script fails half way through for some reason, someone can run it again and only the changes will be applied.  Without using `apply` it's impossible to run this script again because the `create` will fail if the service-account already exists.

brybacki replied:
that is very good point, I'll try to change it now

brybacki replied:
done, it really works great

davidvossel commented on hack/cluster-sync.sh:
```diff
@@ -0,0 +1,56 @@
+# ******************************************************
+./kubevirtci kubectl -n kvcluster create -f ./deploy/infra-cluster-service-account.yaml
+
+export INFRA_KUBECONFIG_IN_TENANT_FILE=_ci-configs/infra_kubeconfig.yaml
```
the `_ci-configs` dir isn't created automatically i don't think.

davidvossel reviewed (commented):
nice! I tried this out and got pretty far. I made some suggestions about how we can evolve this a bit further

nunnatsa commented:
/test pull-csi-driver-linter

nunnatsa commented:
The linter lane fails with timeout. #43 fixes it.

nunnatsa commented:
Trying again. If it will fail, I would merge #43.

/test pull-csi-driver-linter

davidvossel commented:
/retest

nunnatsa commented:
/retest

brybacki commented on hack/cluster-up.sh:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
```
yes

brybacki commented:
I have made files executable, rebased, and updated cluster-sync so it installs kubevirt-csi-driver from local registry. it still uses rook-ceph, that can be changed to local after testing if it works. 

The process is as follows:
```
./hack/cluster-up.sh
export REGISTRY="127.0.0.1:$(cluster-up/cluster-up/cli.sh ports registry)"
make docker-build
make docker-push
./hack/cluster-sync.sh
```
I have tried creating the claim, but it does not work for locally built version. If I use `quay.io/kubevirt/csi-driver:latest` then it works. 

@davidvossel  I did not have the time to check the logs. Wanted to push my Work in Progress so you can use it.

Additional notes: There was one problem with podman push, so I needed to temporarily add --tls-verify=false, but maybe we can correct it and remove the flag. 

I still need to change all the steps to `apply` and use something better for updating templates (`kustomize`)

davidvossel commented:
> I have tried creating the claim, but it does not work for locally built version. If I use quay.io/kubevirt/csi-driver:latest then it works.

I looked into this a bit. I think the issue is that the nested cluster can't access the registry:5000/kubevirt-csi-driver:latest image within the infra. 

```
 - image: registry:5000/kubevirt-csi-driver:latest
    imageID: ""
    lastState: {}
    name: csi-driver
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        message: Back-off pulling image "registry:5000/kubevirt-csi-driver:latest"
        reason: ImagePullBackOff
```

You can see this when looking at the guest cluster's pods

```
[dvossel@localhost kubevirt-csi]$ ./kubevirtci kubectl-tenant get pods -A
Found control plane VM: kvcluster-control-plane-gkzdl in namespace kvcluster
NAMESPACE             NAME                                                    READY   STATUS             RESTARTS   AGE
default               testpodwithcsi                                          0/1     Pending            0          8m52s
kube-system           calico-kube-controllers-5b68c4d876-gj7lz                1/1     Running            0          34m
kube-system           calico-node-tbwts                                       1/1     Running            0          34m
kube-system           calico-node-w2g8n                                       1/1     Running            0          33m
kube-system           coredns-78fcd69978-d4gpl                                1/1     Running            0          34m
kube-system           coredns-78fcd69978-dwfbd                                1/1     Running            0          34m
kube-system           etcd-kvcluster-control-plane-gkzdl                      1/1     Running            0          34m
kube-system           kube-apiserver-kvcluster-control-plane-gkzdl            1/1     Running            0          34m
kube-system           kube-controller-manager-kvcluster-control-plane-gkzdl   1/1     Running            0          34m
kube-system           kube-proxy-6b8hw                                        1/1     Running            0          34m
kube-system           kube-proxy-c68w5                                        1/1     Running            0          33m
kube-system           kube-scheduler-kvcluster-control-plane-gkzdl            1/1     Running            0          34m
kubevirt-csi-driver   kubevirt-csi-controller-d768cc844-qc6tv                 3/4     ImagePullBackOff   0          10m
kubevirt-csi-driver   kubevirt-csi-node-9ct2p                                 2/3     ImagePullBackOff   0          10m
kubevirt-csi-driver   kubevirt-csi-node-zgb9l                                 2/3     ImagePullBackOff   0          10m
```

It's possible that the `registry:5000` dns doesn't work within the guest cluster. When i do a describe on one of those pods i get this event.

```
  Warning  Failed            3s (x2 over 18s)   kubelet            Failed to pull image "registry:5000/kubevirt-csi-driver:latest": rpc error: code = Unknown desc = failed to pull and unpack image "registry:5000/kubevirt-csi-driver:latest": failed to resolve reference "registry:5000/kubevirt-csi-driver:latest": failed to do request: Head "https://registry:5000/v2/kubevirt-csi-driver/manifests/latest": dial tcp: lookup registry: Temporary failure in name resolution
```

I think that registry is typically at ip address `192.168.66.2` though, which is reachable from within the guest cluster. So if I change `registry:5000` to `192.168.66.2:5000` then i get a little further. 

```
Warning  Failed            100s (x2 over 115s)    kubelet            Failed to pull image "192.168.66.2:5000/kubevirt-csi-driver:latest": rpc error: code = Unknown desc = failed to pull and unpack image "192.168.66.2:5000/kubevirt-csi-driver:latest": failed to resolve reference "192.168.66.2:5000/kubevirt-csi-driver:latest": failed to do request: Head "https://192.168.66.2:5000/v2/kubevirt-csi-driver/manifests/latest": http: server gave HTTP response to HTTPS client
```

The error message now changed to HTTP to HTTPs client. 

I think the final step here to resolve this would be to add 192.168.66.2 as an insecure registry to the guest cluster, then things should start working correctly. The guest cluster uses containerd as the runtime, so it probably involves modifying the `/etc/containerd/config.toml` file on the guest nodes. I haven't looked into how this is done, but i'm fairly certain it is possible.

FYI, If you want to debug guest nodes, you can ssh into them using the `kubevirtci ssh-tenant <vm name> <vm namespace>` command. If you want to automate setting some sort of configuration within the guest nodes (like modifying insecure registry configuration), you could probably use a pod with host privileges to do it automatically for you.

brybacki commented:
Yes, the pull command works with insecure registry configured this way:
`/etc/containerd/config.toml`
```
  [plugins."io.containerd.grpc.v1.cri".registry]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors]
      [plugins."io.containerd.grpc.v1.cri".registry.mirrors."192.168.66.2:5000"]
        endpoint = ["http://192.168.66.2:5000"]
    [plugins."io.containerd.grpc.v1.cri".registry.configs]
      [plugins."io.containerd.grpc.v1.cri".registry.configs."192.168.66.2:5000".tls]
        insecure_skip_verify = true
```
`$ sudo crictl pull 192.168.66.2:5000/kubevirt-csi-driver:latest
`

> If you want to automate setting some sort of configuration within the guest nodes (like modifying insecure registry configuration), you could probably use a pod with host privileges to do it automatically for you.

What do you mean by a pod with host privileges? Do we have any examples in kubevirt or other environments?

davidvossel commented:
> What do you mean by a pod with host privileges? Do we have any examples in kubevirt or other environments?

If you just need to alter the `/etc/containerd/config.toml` file, you could probably do that with a pod running as root that mounts `/etc/containerd/config.toml` using a hostPath volume. Here's a link with hostpath examples [1]

In order for the pod to be able to access that config.toml file, it might need to run as root. So you'd need to add something like the following to the pod's yaml.

```
        securityContext:
          runAsUser: 0
```



1. https://www.devopsschool.com/blog/kubernetes-volume-hostpath-explained-with-examples/

awels commented:
So would you not have a problem getting that pod from the registry, since the pod is configuring the registry. Would it make sense to somehow set that in a cloudInit or as part of the templates that are used to build the tenant cluster?

brybacki commented:
> 

I did try some experiments (one of them pushed as "WIP" commit), I have a few steps added to cluster-up, but commented out. When trying manually it works, but it does look like a hack, not the solution. Will try some more ways on monday, maybe it will be better to ssh, run some crio commands? 

@alicefr you mentioned `additionalkubeArgs` or `kubeadm` when we talked? what do you mean?

brybacki commented:
I think, it would be good to have this configured in kubevirtci script in `cluster-api-provider-kubevirt` and if it can be a  `registry:5000`, then it would work for all projects that use this version of kubevirtci in tests.

We could use a single image url on infra and tenant cluster.

davidvossel commented on hack/cluster-up.sh:
```diff
@@ -23,3 +23,9 @@ echo "Enable hotplug"
 
+# TODO: update cluster VMI's with insecure registry...
+# need to run pod per node, and apply change and restart
+# ./kubevirtci kubectl apply -f ./deploy/host-pod.yaml
+# ./kubevirtci kubectl-tenant exec host-pod -- /bin/sh -c "`cat ./deploy/containerd-config.sh`"
```
This is actually a pretty commonly used pattern. Even the openshift conformance tests do this sort of thing.

start a pod somewhere, then exec into it to do something that the tests need. :+1:

davidvossel replied:
just be aware that multiple nodes need this config

brybacki replied:
that is right, a comment above already states it, 
I did test it manually

davidvossel commented:
> When trying manually it works, but it does look like a hack, not the solution. 

This is probably fine. one complication is that you might need this to execute on both the master and worker nodes, which makes it a bit tedious. but still, totally fine if it works.

Another simple solution would be to use a daemonset with an init container that actually does the thing you need to do, and a main container that simply sleeps. The logic would be something like this.

* post a daemonset to the guest cluster with an init container that injects the containerd config and normal container that simply sleeps.
* wait for all daemonsets to come online (we know the init container finished)
* delete daemonset (because we know the config was applied at this point)

> @alicefr you mentioned additionalkubeArgs or kubeadm when we talked? what do you mean?

This sort of thing is possible at guest cluster creation time using the manifests that create the cluster. 

For example, if we look at the kubadm configuration api, there are places we can [inject files](https://github.com/kubernetes-sigs/cluster-api/blob/0017077b17a4d164ca71b7c7d0a5cf64f74b6644/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L54) into the guest cluster during boot strap, and places where we can have [arbitrary  commands execute ](https://github.com/kubernetes-sigs/cluster-api/blob/0017077b17a4d164ca71b7c7d0a5cf64f74b6644/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L66)during bootstrap.

It's this [field in the capk template](https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/blob/43916429f4417c3b18ded7052768e5fb4265c0a6/templates/cluster-template.yaml#L127) that you'd be altering to do something like this.

The alternative is that you inject this sort of thing post cluster creation yourself and reuse the capk templates exactly as they currently are. 

It really doesn't matter much to me as long as you find a way to apply custom host level configuration and it's race free.

alicefr commented:
@brybacki I'm not sure exactly, what you are using. My suggestion was to try to configure crio/containerd through kubelet when you start the cluster. However, I don't see any kubelet flags that pass additional options to the container runtime :/. This will be simpler than creating a privileged pod that reconfigures the container runtime. Additionally, you also need to restart the container runtime if you want that the changes apply. 
Are you using capk? What David suggested with dropping files and executing the command seems relatively easy to me.

davidvossel commented:
I pushed commits to handle the insecure registry logic. I just loop through the vmis and apply the containerd config using ssh. it seems to work reliably

davidvossel reviewed (approved):
/lgtm
/approve

With these changes, we can now run e2e tests by executing the following

```
make cluster-up && make cluster-sync && make e2e-test
```

davidvossel commented on Makefile:
```diff
@@ -45,15 +45,17 @@ include $(addprefix ./vendor/github.com/openshift/build-machinery-go/make/, \
-.PHONY: build-image
-build-image:
+.PHONY: docker-build
+docker-build:
```
i fixed this
